### PR TITLE
[Runtime][Contrib] Support cudnn softmax

### PR DESCRIPTION
**Pull request by icemelon9**
Using cudnn can improve the softmax performance on Nvidia GPU.

@yzhliu @Laurawly @ZihengJiang

### CODE REVIEW

File: src/runtime/contrib/cudnn/softmax.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  int64_t* shape = x->shape;
+  if (axis < 0) axis += ndim;
+  CHECK(axis >= 0 && axis < ndim);
+
+  CuDNNThreadEntry* entry_ptr = CuDNNThreadEntry::ThreadLocal();
+  entry_ptr->softmax_entry.data_type = CuDNNDataType::DLTypeToCuDNNType(x->dtype);
+
+  // Set mode and shape descriptor
+  if (axis == ndim - 1) {
+    int64_t N = 1;
```
Comment (Hzfengsy) [inline src/runtime/contrib/cudnn/softmax.cc:48]: I'm confused why we need `int64_t` here but later cast to `int`

**Comment by tqchen** [comment]
as part of the principle, would be great if we can lookinto making the native op as fast

**Comment by icemelon9** [comment]
@tqchen Yes, I understand that. But the latency could be 10x between tvm schedule and cudnn for the input shape like [100, 1024]. I guess to achieve such performance it requires fusion across multiple stage of reduction, which it seems not easy to be implemented in tir.

File: src/runtime/contrib/cudnn/softmax.cc (hunk 1 above)
Comment (icemelon9) [reply src/runtime/contrib/cudnn/softmax.cc:48]: It's because DLTensor defines the shape in `int64_t` type. There'll be a cast anyway.

**Comment by tqchen** [comment]
k, I am not trying to blocking the PR, merely trying to say it would be great to have such investigation :)

**Comment by tqchen** [comment]
@wpan11nv @yongfeng-nv  can you suggest a bit what are possible optimizations can be done?

**Comment by yongfeng-nv** [comment]
> @wpan11nv @yongfeng-nv can you suggest a bit about possible optimizations that can be done?

We don't know the details, but will look into it.

**Comment by wpan11nv** [comment]
> @wpan11nv @yongfeng-nv can you suggest a bit about possible optimizations that can be done?

The cuda schedule emit 4 kernels, which cause lots of  IO overhead. Ideally, we may emit a single kernel for small reduction sizes  (e.g. reduction dim n <= 1024)

**Pull request merged**